pex/pex_builder.py

```diff
@@ -591,9 +591,11 @@ def _prepare_bootstrap(self):
                 dirs[:] = bootstrap_packages
 
             for f in filter_pyc_files(files):
-                if f.endswith("testing.py"):
-                    continue
                 abs_src = os.path.join(root, f)
+                # N.B.: Some of the `pex.*` package files (__init__.py) will already have been
+                # prepared when vendoring the runtime above; so we skip them here.
+                if abs_src in prepared_sources:
+                    continue
```

Comment:
This particular part regarding files that "will already have been prepared" and how it relates to writing files into the read-only directory is confusing to me and I would appreciate a quick note on how we avoid writing to a read-only dir with this deduplication process if you have a moment. 

This exact line/`if` check is the part that fixes #2203, right? I'm confused as to why not writing duplicate files (which clearly seems good, and probably improves performance too) is fixing the issue with writing to a read-only directory.
- I only see mutating a single `Chroot` instance here, which I assume is a directory we're allowed to write to.
- I see that we reference `_ABS_PEX_PACKAGE_DIR` in this method, which I assume is the bootstrap directory (which may be read-only) that we were previously writing files to, but I don't see how we were writing any files to it.
- I also see that `vendored_sources` stops us from writing the same files to the chroot, which is another deduplication; but that just looks like an optimization (and when I remove it, the new test still passes)?

Thanks for your time, but no pressure to respond, as this is mostly out of curiosity.

Reply:
nvm; found the reasoning at #2203! sorry for ping
> The issue here is that Pex trips over itself when constructing the .bootstrap by attempting to copy pex/__init__.py and other root pex.* package __init__.py files multiple times. The 1st copy works, but since the source file is read only, the destination file is as well, and subsequent copies (over-writes), fail.